Color progress bar depending on build pass/fail state

jjohannes opened the pull request:
This PR implements #2404 as follows:

- Group headers are colored by matching the message `FAILED`: https://github.com/gradle/gradle/commit/05d0447a82f40d8cc685ad4a5f98d4c87ee9c866
- To color the progress bar:
  1. On daemon side: We extend the [ProgressEvent](https://github.com/gradle/gradle/commit/7b58b95937242cc483d39e4d0447c96851a34edd#diff-96938ad51a6f1b6ee358694b93cbecd6) by adding the a [build health](https://github.com/gradle/gradle/commit/7b58b95937242cc483d39e4d0447c96851a34edd#diff-0351f52650c6024be6016425bf189137) state(`UNCHANGED` or `FAILING`). And split the `status` into `status` and `progressIndicator` such that they can be styled individually.
  2. On client side: We style the [progress bar (progressIndicator string)](https://github.com/gradle/gradle/commit/7b58b95937242cc483d39e4d0447c96851a34edd#diff-d4bc723ea2fde4a0e75ac82b6ffc4321) based to the build health

A special case is the `ProgressStartEvent` which also takes in the complete string (progress indicator + status) as `shortDescription` (https://github.com/gradle/gradle/commit/ac43e951b4f1699292670cf061578b66b7ac9289). I think this should eventually not be necessary. It should be enough to get the initial progress through the first `ProgressEvent`. However, removing this breaks a couple of tests and I am not sure if it actually would break functionality. If we want to clean this up, I would propose to do that in a follow up step.

[![asciicast](https://asciinema.org/a/QIJZ4n2bPUYPDk2r3N0bIOeQE.png)](https://asciinema.org/a/QIJZ4n2bPUYPDk2r3N0bIOeQE)

jjohannes commented:
CI Stage 4 Passed: https://builds.gradle.org/viewLog.html?buildId=4722498

eriwen commented on subprojects/core/src/main/java/org/gradle/internal/progress/ProgressFormatter.java:
```diff
@@ -15,7 +15,10 @@
 interface ProgressFormatter {
     String incrementAndGetProgress();
     String getProgress();
+    @Nullable String getProgressIndicator();
```
:x: I think it would be highly preferable not to split a "progress indicator" out here. Instead, we should have `BuildStatusRenderer` manage `ProgressBar`s. 

I had spiked this with other console work [here](https://github.com/gradle/gradle/commit/6115cf15d82b897411a9310ee23a0d04030ab3e6#diff-d4bc723ea2fde4a0e75ac82b6ffc4321), decided to split out the work, and never got back to it.

jjohannes replied:
I have change that such that `BuildProgressLogger` is no longer using any `ProgressFormatter`. Instead the `ProgressEvent` now contains the information about the progress (current, total, failing).
`ProgressBar` is now managed by `BuildStatusRenderer `.

We could probably split up `ProgressEvent`:
(1) events that render on the daemon side and only have a status (as it was before)
(2) events that are rendered on the client side and contain the actual progress information (as we do for the build progress)

Not sure if this has more implications and if we need to do that in this PR. It might not be worth the effort if we want to do more improvements of how build operation progress is rendered in general, as done in your spike, anyway.

eriwen commented on subprojects/logging/src/integTest/groovy/org/gradle/internal/logging/BasicGroupedTaskLoggingFunctionalSpec.groovy:
```diff
@@ -206,6 +206,37 @@ class BasicGroupedTaskLoggingFunctionalSpec extends AbstractConsoleFunctionalSpe
         gradle?.waitForFinish()
     }
 
+    def "group header is printed red if task failed"() {
```
I recommend we break this up and and color the group header in a separate PR.

jjohannes replied:
Moved to #2486

eriwen commented on subprojects/logging/src/main/java/org/gradle/internal/logging/console/BuildStatusRenderer.java:
```diff
@@ -119,10 +132,15 @@ private void phaseStarted(ProgressStartEvent progressStartEvent) {
         currentBuildStatus = progressEvent.getStatus();
+        currentProgressIndicator = progressEvent.getProgressIndicator();
+        if (progressEvent.getBuildHealth() == BuildHealth.FAILING) {
+            currentBuildStatusColor = Style.Color.RED;
```
Is there a way to use `StyledTextOutput.Style.Failure` instead?

jjohannes replied:
`StyledLabel` needs to use `StyledTextOutputEvent.Span`. I can not see why it shouldn't. I changed it: https://github.com/gradle/gradle/pull/2447/commits/f9e9950f9e3c0a428e3ada721a7e2746bdabf057

eriwen commented on subprojects/logging/src/main/java/org/gradle/internal/logging/format/PrettyPrefixedLogHeaderFormatter.java:
```diff
@@ -28,9 +28,14 @@
     }
+
+    private boolean isFailing(@Nullable String status) {
+        return "FAILED".equals(status);
```
:x: It feels a bit fragile to have this logic in the formatter. It seems like the knowledge that the task failed should be an input.

jjohannes replied:
Weill be addressed in #2486

eriwen commented on subprojects/logging/src/main/java/org/gradle/internal/logging/events/BuildHealth.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Indicates the current health of a running build as it progresses.
+ */
+public enum BuildHealth {
```
:o: Some of the terms feel off: `BuildHealth` and `UNCHANGED`. It seems like this could be maintained within `BuildStatusRenderer`

jjohannes replied:
Yes. I think I was overdoing it. I just use a boolean value now.

eriwen commented on subprojects/logging/src/main/java/org/gradle/internal/logging/progress/ProgressLogger.java:
```diff
@@ -110,7 +113,16 @@
+     * @param progressIndicator A printable string that indicates the progress (e.g. [====------])
+     * @param buildHealth The build health.
+     */
+    void progress(@Nullable String status, @Nullable String progressIndicator, BuildHealth buildHealth);
```
I'd like to reiterate how odd it is to split "indicator" and "status". Instead, only `ProgressBar` should have this distinction.

jjohannes replied:
ProgressEvent is still extended now but with the general progress information (see comment above).

jjohannes commented:
Only color the current progress inside the progress bar:

[![asciicast](https://asciinema.org/a/D5M9DZYq9cHjl6sgLwzEMfll0.png)](https://asciinema.org/a/D5M9DZYq9cHjl6sgLwzEMfll0)

jjohannes commented:
Stage 4: https://builds.gradle.org/viewQueued.html?itemId=5546424

jjohannes commented:
I hope I got that right: With _abuse_ the progress logger, you mean that we are sending _totalProgress_ with each progress event instead of sending it once with the start event?

I added another commit which changes that.

I did not introduce a new event as you did in the spike (`PhaseProgressStartEvent`). Instead I added an optional `totalProgress` to the  `ProgressStartEvent`.
I think the events should rather be composed of optional values than using subclassing. Everything will be flattened by the serialisation. So I think we should prefer composition over subclassing here. And it seems to me like that is the pattern that has been used before. E.g. `ProgressStartEvent` already contains a bunch of optional values.

In the end the changes to the progress logging API here just make information explicit (total progress, failed) that was encoded in the status String before. So for my understanding these are valid extensions.

Do I understand it right that what you are heading for with the spike is making even more information available on the client side. Eventually transporting all build operation details to the client?
Then we could indeed have a slim API that just transports any kind of build operation and then we make decisions about what actually is the progress on the client side (e.g. by looking at `BuildOperationCategory.TASK`). I feel that should be a separate piece of work though.

jjohannes commented:
> Aside, one thing we don't yet consider is test failures. For example, we don't turn the bar red if there's a test failure until the Test task fails.

We also don't show progress for single tests completion. Which would also be a very nice improvement. At the moment the total progress `hangs` for long running test tasks. Like this

`<============-> 99% EXECUTING [46s]`

1 minute later:

`<============-> 99% EXECUTING [1m 46s]`

Feels like these two issues are strongly connected. Should we open a new issue to address both?

jjohannes commented:
> In a separate PR it'd be cool if we could keep the ProgressBar on the screen for the build summary.

I created #2511

jjohannes commented:
Ad discussed with @eriwen I am merging this. I did some more manual testing and a CI build (https://builds.gradle.org/viewLog.html?buildId=6058022). No issues came up.

For following up on testing progress, I created #2543.
